man/vdevmap.8
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.41.2.
.TH VDEVMAP "8" "May 2013" "vdevmap  -- manages device mappings" "System Administration"
.SH NAME
vdevmap \- manages device mappings
.SH SYNOPSIS
.B vdevmap
\fI--xid <xid> {--set|--unset} \fR[\fI--flags <flags>\fR] [\fI--open\fR] [\fI--create\fR]
.SH DESCRIPTION
.IP
[\-\-device <dev>] [\-\-remap \fB\-\-target\fR <dev>]
.TP
\fB\-\-flags\fR <flags>
Set the specified flags
.TP
\fB\-\-open\fR
Allow opening of the device
.TP
\fB\-\-create\fR
If CAP_MKNOD is given, allow mknod(2)
.TP
\fB\-\-device\fR <dev>
Device to apply the command to
.TP
\fB\-\-remap\fR
Remap the device to the target
.TP
\fB\-\-target\fR <dev>
Target for \fB\-\-remap\fR
.PP
EXAMPLES
.IP
Remap \fI/dev/hda1\fP to \fI/dev/vroot1\fP for xid 42
.IP
vdevmap \fB\-\-xid\fR 42 \fB\-\-set\fR \fB\-\-open\fR \fB\-\-device\fR \fI/dev/hda1\fP \fB\-\-target\fR \fI/dev/vroot1\fP \fB\-\-remap\fR
.IP
Let xid 42 create all device nodes
.IP
vdevmap \fB\-\-xid\fR 42 \fB\-\-set\fR \fB\-\-open\fR \fB\-\-create\fR \fB\-\-target\fR \fI/dev/null\fP
vdevmap \fB\-\-xid\fR 42 \fB\-\-set\fR \fB\-\-open\fR \fB\-\-create\fR \fB\-\-target\fR \fI/dev/root\fP
.IP
Let xid 43 create just \fI/dev/null\fP
.IP
vdevmap \fB\-\-xid\fR 43 \fB\-\-set\fR \fB\-\-open\fR \fB\-\-create\fR \fB\-\-device\fR \fI/dev/null\fP
.PP
Please report bugs to vserver@list.linux\-vserver.org
.PP
This program is part of util\-vserver
.SH COPYRIGHT
Copyright \(co 2006 Daniel Hokka Zakrisson
This program is free software; you may redistribute it under the terms of
the GNU General Public License.  This program has absolutely no warranty.


NEWS
version 0.30.215
================

	- added support for using libnss instead of beecrypt for
	vhashify's hash calculation.  libnss has bad SHA1 performance on
	i386, but is a) maintained and b) gives better performance with
	all other hashes and on x86_64.

	- added tunctl, a small program to setup tun/tap interfaces, with
	support in the configuration.

	- vattribute and nattribute now support a --get operation, which
	parses the flags and capabilities into a human-readable form.

	- vspace supports all the spaces currently available.

	- vmemctrl controls the OOM bias of context.

	- dynamic contexts are unbroken.

	- hopefully all rpm-fake races have been fixed.

	- flock has been replaced by lockf, which works on NFS.

	- the clone build method now supports exclude lists.

	- vmount handles mounting into a guest.

	- various fixes for the Gentoo-scripts, thanks to Benedikt Böhm.

	- support for yum's persistdir has been added.

	- per-guest device maps are now supported.

	- vserver ... build now sets the barrier on the parent directory.

	- fsck is now supported for the guest's fstab.

	- symlinks are now handled on all levels in namespace cleanup.

	- the debootstrap build method now supports debootstrap 1.0.8.

	- a documentation update, thanks to Micah Anderson.

	- use numeric uids/gids for the template build method.

	- add support for the alpine and f8 distributions.
	
version 0.30.214
================
	- patches for yum-3.2.

	- various fixes to allow for *SuSE guests.

	- added Fedora 7.

	- replace file with h2ext for build -m template.

	- add exclude list support for vclone.

	- include debootstrap scripts for a lot of distributions.

	- fixes to allow the Gentoo initscripts to be used with baselayout 2.

	- remove disk limit caches on vserver ... delete.

	- added support for the new APIs in 2.2 and 2.3, in particular the
	netv2 API, allowing native IPv6 support.

	- added support for a separate filesystem tag.

version 0.30.213
================
	- added support for the device mapping feature present in development
	kernels.

	- further improved Gentoo support, again thanks to Benedikt Böhm.

	- support persistent contexts.

	- add another initscript that creates directories, configures vshelper
	as well as kills all contexts.

	- new vserver-stat implementation, which uses the accounting APIs
	available in Linux-VServer 2.2.

	- added support for setting sysctl values in the guest configuration.

	- added a clone build method, which simply links already unified files.

	- implement dynamic contexts in userspace, generating a static one
	during vserver ... build.

	- fix the endless loop in vlogin.

	- fix Ubuntu builds.

	- improved support for Debian Etch, thanks to Benedikt Böhm.

	- added CentOS 5.

	- use the state change helper to try to cleanup after us if the guest
	exited on its own.

version 0.30.212
================
	- support for the Linux-VServer 2.1 API has been implemented.

	- the scheduler configuration has been split up into multiple files.

	- namespace cleanup is enabled by default. This can be disabled by
	touch ${sysconfdir}/vservers/.defaults/nonamespace-cleanup

	- all unset ulimits are now reset to infinity, so limits are no longer
	inherited from the host.

	- much improved Gentoo support, thanks to Benedikt Böhm.

version 0.30.211
================

	- novlandev is now the default. To reenable the old behaviour of
	creating VLAN interfaces where used, touch
	${sysconfdir}/vservers/.defaults/interfaces/vlandev.

	- vshelper now works correctly with the kernel legacy support turned
	off.

	- vlogin provides a terminal proxy which allocates a new psuedo-tty
	on the inside when you enter a guest.

	- and lots of other features/fixes, see ChangeLog.


version 0.30.210
================

	- LEGACY BREAKAGE: by default, only recent kernel APIs will be
	supported so you have to make sure to use at least the vs2.0.1
	kernel patch. With older kernels, you have to call 'configure'
	with '--enable-apis=NOLEGACY'. Some programs will not work anymore
	with the default settings; these ones are:

	     + capchroot
	     * chcontext-context
	     * reducecap

	- dlimit support was integrated into the standard vserver
	operations. Thanks to Daniel Hokka Zakrisson for his patch.

	- recent kernel API is now used for setting up the network
	contexts. So it should be now possible to use a kernel compiled
	without CONFIG_VSERVER_LEGACYNET. Thanks to Daniel Hokka Zakrisson
	for his patch.  .


version 0.30.209
================

	- large revamping...
	
	
version 0.23.6
==============

	- fixed '--level' option on 'vserver ... chkconfig'


src/vdevmap.c
// $Id$    --*- c -*--

// Copyright (C) 2006 Daniel Hokka Zakrisson
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include "util.h"
#include <lib/internal.h>

#include <vserver.h>

#include <getopt.h>
#include <errno.h>

#define ENSC_WRAPPERS_PREFIX	"vdevmap: "
#define ENSC_WRAPPERS_UNISTD	1
#define ENSC_WRAPPERS_VSERVER	1
#include <wrappers.h>

#define CMD_HELP		0x1000
#define CMD_VERSION		0x1001

int		wrapper_exit_code  =  1;

struct option const
CMDLINE_OPTIONS[] = {
  { "help",       no_argument,       0, CMD_HELP },
  { "version",    no_argument,       0, CMD_VERSION },
  { "xid",        required_argument, 0, 'x' },
  { "set",        no_argument,       0, 's' },
  { "unset",      no_argument,       0, 'u' },
  { "open",       no_argument,       0, 'o' },
  { "create",     no_argument,       0, 'c' },
  { "remap",      no_argument,       0, 'r' },
  { "flags",      required_argument, 0, 'f' },
  { "device",     required_argument, 0, 'd' },
  { "target",     required_argument, 0, 't' },
  {0,0,0,0}
};

static void
showHelp(int fd, char const *cmd)
{
  WRITE_MSG(fd, "Usage: ");
  WRITE_STR(fd, cmd);
  WRITE_MSG(fd,
	    " --xid <xid> {--set|--unset} [--flags <flags>] [--open] [--create]\n"
	    "        [--device <dev>] [--remap --target <dev>]\n"
	    "\n"
	    "    --flags <flags>     Set the specified flags\n"
	    "    --open              Allow opening of the device\n"
	    "    --create            If CAP_MKNOD is given, allow mknod(2)\n"
	    "    --device <dev>      Device to apply the command to\n"
	    "    --remap             Remap the device to the target\n"
	    "    --target <dev>      Target for --remap\n"
	    "\n"
	    "EXAMPLES\n"
	    "  Remap /dev/hda1 to /dev/vroot1 for xid 42\n"
	    "    vdevmap --xid 42 --set --open --device /dev/hda1 --target /dev/vroot1 --remap\n"
	    "  Let xid 42 create all device nodes\n"
	    "    vdevmap --xid 42 --set --open --create --target /dev/null\n"
	    "    vdevmap --xid 42 --set --open --create --target /dev/root\n"
	    "  Let xid 43 create just /dev/null\n"
	    "    vdevmap --xid 43 --set --open --create --device /dev/null\n"
	    "\n"
	    "Please report bugs to " PACKAGE_BUGREPORT "\n");

  exit(0);
}

static void
showVersion()
{
  WRITE_MSG(1,
	    "vdevmap " VERSION " -- manages device mappings\n"
	    "This program is part of " PACKAGE_STRING "\n\n"
	    "Copyright (C) 2006 Daniel Hokka Zakrisson\n"
	    VERSION_COPYRIGHT_DISCLAIMER);
  exit(0);
}

int main(int argc, char *argv[])
{
  xid_t		xid		= VC_NOCTX;
  bool		allow_open	= false;
  bool		allow_create	= false;
  bool		do_remap	= false;
  uint32_t	flags		= 0;
  char		*device		= NULL;
  char		*target		= NULL;
  bool		set		= true;
  unsigned long	tmp		= 0;
  
  while (1) {
    int		c = getopt_long(argc, argv, "+x:suocrf:d:t:", CMDLINE_OPTIONS, 0);
    if (c==-1) break;

    switch (c) {
      case CMD_HELP	:  showHelp(1, argv[0]);
      case CMD_VERSION	:  showVersion();
      case 'x'		:  xid = Evc_xidopt2xid(optarg, true);	break;
      case 'o'		:  allow_open = true;			break;
      case 'c'		:  allow_create = true;			break;
      case 'r'		:  do_remap = true;			break;
      case 'd'		:  device = optarg;			break;
      case 't'		:  target = optarg;			break;
      case 's'		:  set = 1;				break;
      case 'u'		:  set = 0;				break;
      case 'f'		:
	if (!isNumberUnsigned(optarg, &tmp, false)) {
	  WRITE_MSG(2, "Invalid flags argument: '");
	  WRITE_STR(2, optarg);
	  WRITE_MSG(2, "'; try '--help' for more information\n");
	  return EXIT_FAILURE;
	}
	flags |= (uint32_t) tmp;
	break;

      default		:
	WRITE_MSG(2, "Try '");
	WRITE_STR(2, argv[0]);
	WRITE_MSG(2, " --help' for more information.\n");
	return EXIT_FAILURE;
	break;
    }
  }

  if (allow_open)	flags |= VC_DATTR_OPEN;
  if (allow_create)	flags |= VC_DATTR_CREATE;
  if (do_remap)		flags |= VC_DATTR_REMAP;

  if (!target && do_remap)
    WRITE_MSG(2, "Remapping specified without a target; try '--help' for more information\n");
  else if (xid==VC_NOCTX)
    WRITE_MSG(2, "No xid specified; try '--help' for more information\n");
  else if (optind!=argc)
    WRITE_MSG(2, "Unused argument(s); try '--help' for more information\n");
  else if (!device && !target)
    WRITE_MSG(2, "Device and target are missing; try '--help' for more information\n");
  else if (set && vc_set_mapping(xid, device, target, flags)==-1)
      perror("vc_set_mapping()");
  else if (!set && vc_unset_mapping(xid, device, target, flags)==-1)
      perror("vc_unset_mapping()");
  else
    return EXIT_SUCCESS;

  return EXIT_FAILURE;
}
